hw3/README.md

```diff
@@ -0,0 +1,376 @@
+assume you're extending a reasonably-well-structured existing parser.
+
+To demonstrate that your parser works, you must to implement the
+parse-only mode of the [command line specification][jpl-p]. In this
+mode, your compiler should parse the input file and produce an output
+that looks like this:
+
+[jpl-p]: https://github.com/utah-cs4470-sp23/class/blob/main/spec.md
+
+```
+(ReadCmd "photo.png" (VarArgument photo))
+(AssertCmd (VarExpr a) "Photo must be 800 pixels wide")
+(AssertCmd (VarExpr b) "Photo must be 600 pixels tall")
+(LetCmd (ArgLValue (VarArgument middle)) (VarExpr photo))
+(WriteCmd (VarExpr middle) "profile.png")
+```
+
+Your compiler must also print `Compilation failed` when a parser error
+is encountered and `Compilation succeeded` when parsing works. We will
+use this when testing your code. The output format is formally defined
+below.
+
+## Output format
+
+Your parser should produce output in what's called an "S-expression"
+format. This format can be generated directly from your AST nodes. We
+recommend giving each AST node a `toString` method (or your language's
+equivalent) that turns that AST node into the S-expression string.
+
+While S-expressions look kind of confusing, they follow a very simple
+rule: every AST node should turn into:
+
+- An open parenthesis
+- Followed by the name of the AST node
+- Followed by each field of the AST node
+  - each preceded by a space
+  - in the same order as in the grammar
+  - if it is an AST node field, print its S-expression
+  - if it is a `<string>` field, print it with quotation marks
+  - otherwise, just print it directly
+- Followed by a close parenthesis
+
+For example, the grammatical rule for `read` is:
+
+    expr : read image <string> to <argument>
+
+The corresponding AST node is called `ReadCmd` and has two fields, a
+string and an argument. So its `toString` method would look like this:
+
+```
+public class ReadCmd extends Cmd {
+    String filename;
+    Argument arg;
+
+    public String toString() {
+        out = "(ReadCmd";
+        out += " " + this.string_sexp(this.filename);
+        out += this.arg.toString();
+        out += ")";
+        return out;
+    }
+}
+```
+
+Here the `string_sexp` method takes a `String` argument and returns
+that string with double quotes around it:
+
+```
+public class ASTNode {
+
+    public String string_sexp(String input) {
+        return '"' + input + '"';
+    }
+}
+```
+
+Note that in this implementation, the `ReadCmd` calls `toString` on an
+`Argument`. That would also implement a `toString` method, like this:
+
+```
+public class VarArgument extends Argument {
+    String name;
+
+    public String toString() {
+        out = "(VarArgument";
+        out += " " + this.name;
+        out += ")";
+        return out;
+    }
+}
+```
+
+The AST node names that you must use in your S-expression output are:
+
+    ReadCmd WriteCmd TypeCmd LetCmd AssertCmd PrintCmd ShowCmd
+    IntType FloatType BoolType
```

Comment:
Should "VarType" be included in this list?

Reply:
Good catch